pkg/cloud/services/networking/securitygroups.go

```diff
@@ -399,6 +401,38 @@ func (s *Service) createRule(r infrav1.SecurityGroupRule) (infrav1.SecurityGroup
 	return convertOSSecGroupRuleToConfigSecGroupRule(*rule), nil
 }
 
+func (s *Service) ReconcileSecurityGroupsToInstanceAttachedPorts(openStackMachine *infrav1.OpenStackMachine, attachedPorts []string, desiredSecurityGroupIDs []string) error {
+	for _, portID := range attachedPorts {
+		portUpdateNeeded := false
+		port, err := s.client.GetPort(portID)
+		if err != nil {
+			return fmt.Errorf("error getting port %s: %v", portID, err)
+		}
+
```

Comment:
Instead of doing https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1774/commits/1593e43e66b1f8a912c158e6df15b354f6d7cc10 - we could:

* Put a new struct which maps ports and their security group IDs in OpenStackMachine.Status so we do the lookup once
* we don't want to add these SG IDs to the slices in the openstackmachine_controller that we created for SG IDs to reconcile because they are specific to ports and not applied for all ports (unlike ManagedSecurityGroups & Spec.SecurityGroups)
* Here in the reconcile, if a port is found in the Status with SG IDs, we simply add them to the desired SGs. 

This is a bit complex but this is to address the 3 current options that we have to attach security groups to machines:
* OpenStackCluster.Spec.ManagedSecurityGroups
* OpenStackMachine.Spec.SecurityGroups
* OpenStack.Machine.Spec.Ports.SecurityGroupFilters

Reply:
Another option is we don't reconcile the security groups from ports defined in `OpenStack.Machine.Spec.Ports` because we consider them immutable. That feels odd to me, and I can't find yet a good option here that doesn't bring complexity.